source/connection_manager.c

```diff
@@ -801,9 +794,69 @@ static void s_schedule_connection_culling(struct aws_http_connection_manager *ma
 
-    aws_event_loop_schedule_task_future(manager->cull_event_loop, manager->cull_task, cull_task_time);
+    return cull_task_time;
+}
+
+static uint64_t s_calculate_pending_connections_acquire_cull_task_time(struct aws_http_connection_manager *manager) {
+    if (manager->pending_connections_acquire_timeout_ms == 0) {
+        return 0;
+    }
+
+    uint64_t cull_task_time = 0;
+    aws_mutex_lock(&manager->lock);
+
+    const struct aws_linked_list_node *end = aws_linked_list_end(&manager->pending_acquisitions);
+    struct aws_linked_list_node *oldest_node = aws_linked_list_begin(&manager->pending_acquisitions);
+    if (oldest_node != end) {
+        /*
+         * Since the pending acquires are in FIFO order in the list, the front of the list has the closest
+         * cull time.
+         */
+        struct aws_http_connection_acquisition *oldest_pending_acquire =
+            AWS_CONTAINER_OF(oldest_node, struct aws_http_connection_acquisition, node);
+        cull_task_time = oldest_pending_acquire->timeout_timestamp;
+    } else {
+        /*
+         * There are no connections in the list, so the absolute minimum anything could be culled is the full
+         * culling interval from now.
+         */
+        uint64_t now = 0;
+        manager->system_vtable->aws_high_res_clock_get_ticks(&now);
+        cull_task_time =
+            now + aws_timestamp_convert(
+                      manager->pending_connections_acquire_timeout_ms, AWS_TIMESTAMP_MILLIS, AWS_TIMESTAMP_NANOS, NULL);
+    }
+    aws_mutex_unlock(&manager->lock);
+    return cull_task_time;
+}
+
+static void s_schedule_culling(struct aws_http_connection_manager *manager) {
+    if (manager->max_connection_idle_in_milliseconds == 0 && manager->pending_connections_acquire_timeout_ms == 0) {
+        return;
+    }
+
+    if (manager->cull_task == NULL) {
+        manager->cull_task = aws_mem_calloc(manager->allocator, 1, sizeof(struct aws_task));
+        aws_task_init(manager->cull_task, s_cull_task, manager, "cull_idle_connections");
+        /* For the task to properly run and cancel, we need to keep manager alive */
+        aws_ref_count_acquire(&manager->internal_ref_count);
+    }
+
+    if (manager->cull_event_loop == NULL) {
+        manager->cull_event_loop = aws_event_loop_group_get_next_loop(manager->bootstrap->event_loop_group);
+    }
+    AWS_FATAL_ASSERT(manager->cull_event_loop != NULL);
+
+    uint64_t cull_task_time = s_calculate_idle_connection_cull_task_time(manager);
+    uint64_t connection_acquire_timeout = s_calculate_pending_connections_acquire_cull_task_time(manager);
 
-    return;
+    if (manager->max_connection_idle_in_milliseconds != 0 && manager->pending_connections_acquire_timeout_ms != 0) {
+        cull_task_time = aws_min_u64(cull_task_time, connection_acquire_timeout);
+    } else if (manager->pending_connections_acquire_timeout_ms != 0) {
+        cull_task_time = connection_acquire_timeout;
+    }
```

Comment:
do we need two separate helper function to get the proper time for next cull task?
it seems more complicated than it needs to be.

Reply:
I started with one function, but it was more confusing since we need to cast the list node to different structs and check a different variable for full cull timestamp. It was the same amount of code duplication. So I split them into two separate functions.